control/diagnostics_tool/src/can_monitor.c

```diff
@@ -0,0 +1,335 @@
+/************************************************************************/
+/* The MIT License (MIT) */
+/* ===================== */
+
+/* Copyright (c) 2016 PolySync Technologies, Inc.  All Rights Reserved. */
+
+/* Permission is hereby granted, free of charge, to any person */
+/* obtaining a copy of this software and associated documentation */
+/* files (the “Software”), to deal in the Software without */
+/* restriction, including without limitation the rights to use, */
+/* copy, modify, merge, publish, distribute, sublicense, and/or sell */
+/* copies of the Software, and to permit persons to whom the */
+/* Software is furnished to do so, subject to the following */
+/* conditions: */
+
+/* The above copyright notice and this permission notice shall be */
+/* included in all copies or substantial portions of the Software. */
+
+/* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES */
+/* OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND */
+/* NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT */
+/* HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, */
+/* WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING */
+/* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR */
+/* OTHER DEALINGS IN THE SOFTWARE. */
+/************************************************************************/
+
+/**
+ * @file can_monitor.c
+ * @brief Can monitor interface source.
+ *
+ */
+
+
+
+
+#include <sys/time.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "macros.h"
+#include "terminal_print.h"
+#include "can_monitor.h"
+
+
+
+
+// *****************************************************
+// static global data
+// *****************************************************
+
+
+//
+static can_frame_s can_msg_array[ CAN_MSG_ARRAY_SIZE ];
+
+
+//
+static struct timeval tv;
+
+
+
+
+// *****************************************************
+// static definitions
+// *****************************************************
+
+
+//
+static int update_can_array_contents(
+        const unsigned int msg_dlc,
+        const unsigned int msg_flag,
+        const unsigned char * const buffer,
+        can_frame_contents_s * const can_frame_contents )
+{
+    int i;
+
+    int msg_contents_changed = 0;
+
+    can_frame_contents->msg_dlc = msg_dlc;
+
+    can_frame_contents->msg_flag = msg_flag;
+
+    for( i = 0; i < 8; i ++ )
```

Comment:
What about using if(!memcmp(...)) and memcpy here instead of the nested for loops?